src/core/connection.cpp

```diff
@@ -33,360 +33,631 @@
+     *
+     *    const unsigned MAX_NUM_REDIRECTIONS = 3;
+     *
+     *    for (unsigned num_redirections = 0; num_redirections < MAX_NUM_REDIRECTIONS;
+     *         ++num_redirections) {
+     *        p_websocket->set_receive_timeout(std::chrono::seconds(10));
+     *
+     *        const std::string uri = p_websocket->uri();
+     *
+     *        std::unique_ptr<Connection> p(new Connection(std::move(p_websocket), std::move(p_error_handler)));
+     *
+     *        Buffer buffer;
+     *        parser::MessageList messages;
+     *        ConnectionState& state = p->state_;
+     *
+     *        if (p->receive_(&buffer, &messages) != websockets::State::OPEN)
+     *            return p;
+     *
+     *        if (messages.size() != 1 || state != ConnectionState::CHALLENGING) {
+     *            p->close();
+     *            return p;
+     *        }
+     *
+     *        {
+     *            MessageBuilder chr(Topic::CONNECTION, Action::CHALLENGE_RESPONSE);
+     *            chr.add_argument(uri);
+     *
+     *            if (p->send_(chr) != websockets::State::OPEN)
+     *                return p;
+     *        }
+     *
+     *        if (p->receive_(&buffer, &messages) != websockets::State::OPEN)
+     *            return p;
+     *
+     *        if (messages.size() != 1) {
+     *            p->close();
+     *            return p;
+     *        }
+     *
+     *        if (state == ConnectionState::AWAIT_AUTHENTICATION)
+     *            return p;
+     *
+     *        if (state == ConnectionState::AWAIT_CONNECTION) {
+     *            const Message& redirect_msg = messages.front();
+     *            assert(redirect_msg.num_arguments() == 1);
+     *
+     *            const Buffer& uri_buffer = redirect_msg[0];
+     *            std::string new_uri(uri_buffer.cbegin(), uri_buffer.cend());
+     *
+     *            p_websocket = p->p_websocket_->construct(new_uri);
+     *            p_error_handler = std::move(p->p_error_handler_);
+     *
+     *            continue;
+     *        }
+     *
+     *        p->close();
+     *        return p;
+     *    }
+     *
+     *    assert(p_error_handler);
+     *    p_error_handler->onError(ErrorState::TOO_MANY_REDIRECTIONS);
+     *
+     *    return std::unique_ptr<Connection>(
+     *        new Connection(nullptr, std::move(p_error_handler)));
+     *}
+     *
+     *std::unique_ptr<Connection> Connection::make(const std::string& uri,
+     *    std::unique_ptr<ErrorHandler> p_eh, WSHandler &ws_handler)
+     *{
+     *    if (uri.empty())
+     *        throw std::invalid_argument("URI must not be empty");
+     *
+     *    //return Client::make(std::unique_ptr<websockets::Client>(websockets::poco::Client::makeClient(uri)), std::move(p_eh));
+     *    return nullptr;
+     *}
+     */
+
+    void Connection::login(const std::string& auth_params, const Client::LoginCallback &callback)
+    {
+        //if (state_ != ConnectionState::AWAIT_AUTHENTICATION) {
```

Comment:
we should check if already logged in, just witnessed someone trying logging in multiple times